From 3ac579a0844c200b71e160dd49208bc53ce06e8d Mon Sep 17 00:00:00 2001
From: Ondrej Holy <oholy@redhat.com>
Date: Tue, 28 Apr 2026 18:37:44 +0000
Subject: [PATCH] [client,X11] fix residual race in xf_clipboard_formats_free

Backport of commit 4c9f7e8a7129c8be15f6e2686559d3f17936677d.

Adjusted hunk offsets for 3.10.3, `nullptr` replaced with `NULL`.

Made-with: Cursor
---
 client/X11/xf_cliprdr.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/client/X11/xf_cliprdr.c b/client/X11/xf_cliprdr.c
index 7a8ef71..82878e7 100644
--- a/client/X11/xf_cliprdr.c
+++ b/client/X11/xf_cliprdr.c
@@ -850,13 +850,14 @@ static void xf_clipboard_formats_free(xfClipboard* clipboard)
 {
 	WINPR_ASSERT(clipboard);
 
-	/* Synchronize RDP/X11 thread with channel thread */
+	/* Synchronize RDP/X11 thread with channel thread.
+	 * Reset the pointer and count inside the lock so that
+	 * xf_clipboard_changed cannot observe a freed-but-non-NULL pointer. */
 	xf_lock_x11(clipboard->xfc);
 	xf_cliprdr_free_formats(clipboard->lastSentFormats, clipboard->lastSentNumFormats);
-	xf_unlock_x11(clipboard->xfc);
-
 	clipboard->lastSentFormats = NULL;
 	clipboard->lastSentNumFormats = 0;
+	xf_unlock_x11(clipboard->xfc);
 }
 
 static BOOL xf_clipboard_copy_formats(xfClipboard* clipboard, const CLIPRDR_FORMAT* formats,
-- 
2.53.0

